src/smirnoffio.py

```diff
@@ -427,11 +456,44 @@ def __init__(self,options,tgt_opts,forcefield):
         self.set_option(tgt_opts,'mol2',forceprint=True)
         self.set_option(tgt_opts,'coords',default="all.gro")
         self.set_option(tgt_opts,'openmm_precision','precision',default="double", forceprint=True)
-        self.set_option(tgt_opts,'openmm_platform','platname',default="CUDA", forceprint=True)
+        self.set_option(tgt_opts,'openmm_platform','platname',default="Reference", forceprint=True)
         self.engine_ = SMIRNOFF
         ## Initialize base class.
         super(AbInitio_SMIRNOFF,self).__init__(options,tgt_opts,forcefield)
 
+    def submit_jobs(self, mvals, AGrad=False, AHess=False):
+        # we update the self.pgrads here so it's not overwritten in rtarget.py
+        self.smirnoff_update_pgrads()
+
+    def smirnoff_update_pgrads(self):
+        """
+        Update self.pgrads based on smirks present in mol2 files
+
+        This can greatly improve gradients evaluation in big optimizations
+
+        Note
+        ----
+        1. This function assumes the names of the forcefield parameters has the smirks as the last item
+        2. This function assumes params only affect the smirks of its own. This might not be true if parameter_eval is used.
+        """
+        orig_pgrad_set = set(self.pgrad)
+        # smirks to param_idxs map
+        smirks_params_map = defaultdict(list)
+        for pname, pidx in self.FF.map.items():
+            smirks = pname.rsplit('/',maxsplit=1)[-1]
+            smirks_params_map[smirks].append(pidx)
```

Comment:
Note: here the `smirks_params_map` is built by parsing parameter names. When `parameter_eval` or `parameter_repeat` is used, this may give an incomplete map, i.e. one parameter may affect multiple SMIRKs when it is used in the `parameter_eval` of other parameters that belongs to other SMIRKs.
A potential way to improve this is to build an explicit mapping of SMIRKs -> parameters, when parsing the forcefield file.